ompi/mpi/c/type_create_darray.c.in

```diff
@@ -0,0 +1,104 @@
+                                         INT ndims,
+                                         COUNT_ARRAY gsize_array,
+                                         INT_ARRAY distrib_array,
+                                         INT_ARRAY darg_array,
+                                         INT_ARRAY psize_array,
+                                         INT order,
+                                         DATATYPE oldtype,
+                                         DATATYPE_OUT newtype)
+{
+    int i, rc;
+
+    MEMCHECKER(
+        memchecker_datatype(oldtype);
+    );
+
+    if (MPI_PARAM_CHECK) {
+        int prod_psize = 1;
+        OMPI_ERR_INIT_FINALIZE(FUNC_NAME);
+        if( (rank < 0) || (size < 0) || (rank >= size) ) {
+            return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+        } else if( ndims < 0 ) {
+            return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_COUNT, FUNC_NAME);
+        } else if( (ndims > 0) && ((NULL == gsize_array) || (NULL == distrib_array) ||
+                                   (NULL == darg_array) || (NULL == psize_array))) {
+            return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+        } else if (NULL == newtype) {
+            return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_TYPE, FUNC_NAME);
+        } else if( !(OPAL_DATATYPE_FLAG_DATA & oldtype->super.flags) ) {
+            return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_TYPE, FUNC_NAME);
+        } else if( (MPI_ORDER_C != order) && (MPI_ORDER_FORTRAN != order) ) {
+            return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+        }
+        if( ndims > 0 ) {
+            for( i = 0; i < ndims; i++ ) {
+                if( (MPI_DISTRIBUTE_BLOCK != distrib_array[i]) &&
+                    (MPI_DISTRIBUTE_CYCLIC != distrib_array[i]) &&
+                    (MPI_DISTRIBUTE_NONE != distrib_array[i]) ) {
+                    return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+                } else if( (gsize_array[i] < 1) || (psize_array[i] < 0) ||
+                           ((darg_array[i] < 0) && (MPI_DISTRIBUTE_DFLT_DARG != darg_array[i]) ) ) {
+                    return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+                } else if( (MPI_DISTRIBUTE_DFLT_DARG != darg_array[i]) &&
+                           (MPI_DISTRIBUTE_BLOCK == distrib_array[i]) &&
+                           ((darg_array[i] * psize_array[i]) < gsize_array[i]) ) {
+                    return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+                } else if( 1 > psize_array[i] )
+                    return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+                prod_psize *= psize_array[i];
+            }
+            if( prod_psize != size )
+                return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_ARG, FUNC_NAME);
+        }
+    }
+
+    rc = ompi_datatype_create_darray( size, rank, ndims,
```

Comment:
This line creates a compiler warning about mismatched types.

Reply:
fixed with a temporary patch to be removed once we embiggen the datatype engine.